rmqproducer/rabbitmq_producer.py

```diff
@@ -0,0 +1,390 @@
+            self._connection = connection
+        else:
+            self._LOGGER.info('Connection received from connection pool')
+            self._connection = connection
+            self.add_on_connection_close_callback()
+            self.open_channel()
+
+    def on_connection_open(self, unused_connection):
+        """This method is called by pika once the connection to RabbitMQ has
+        been established.
+
+        It passes the handle to the connection object in case we need it, but
+        in this case, we'll just mark it unused.
+
+        :type unused_connection: pika.SelectConnection
+
+        """
+        self._LOGGER.info('Connection opened')
+        self.add_on_connection_close_callback()
+        self.open_channel()
+
+    def add_on_connection_close_callback(self):
+        """This method adds an on close callback that will be invoked by pika
+        when RabbitMQ closes the connection to the publisher unexpectedly.
+
+        """
+        self._LOGGER.info('Adding connection close callback')
+        self._connection.add_on_close_callback(self.on_connection_closed)
+
+    def on_connection_closed(self, connection, reply_code, reply_text):
+        """This method is invoked by pika when the connection to RabbitMQ is
+        closed unexpectedly. Since it is unexpected, we will reconnect to
+        RabbitMQ if it disconnects.
+
+        :param pika.connection.Connection connection: The closed connection obj
+        :param int reply_code: The server provided reply_code if given
+        :param str reply_text: The server provided reply_text if given
+
+        """
+        self._channel = None
+        if self._connection_closing:
+            self._LOGGER.info('Connection was closed: (%s) %s',
+                              reply_code, reply_text)
+            self._connection.ioloop.stop()
+        else:
+            self._LOGGER.warning('Connection closed, reopening in %d seconds: (%s) %s',
+                                 self.reconnect_time, reply_code, reply_text)
+            self._connection.add_timeout(self.reconnect_time, self.reconnect)
+
+    def reconnect(self):
+        """Will be invoked by the IOLoop timer if the connection is
+        closed. See the on_connection_closed method.
+
+        """
+        unpublished_messages = self._messages
+        self.reset_messages()
+
+        # This is the old connection IOLoop instance, stop its ioloop
+        self._connection.ioloop.stop()
+        RMQConnectionPool.remove_connection(self._url)
+
+        # Create a new connection
+        self.connect()
+
+        # There is now a new connection, needs a new ioloop to run
+        self._connection.ioloop.start()
+        if len(unpublished_messages) > 0:
+            self._LOGGER.info("Publishing Messages left on reconnection")
+            for item in unpublished_messages.values():
+                self.publish_message(item['message'], item['routing_key'])
```

Comment:
I mean the failed message number will be different compared to the retry message number!!